java-checks/src/main/java/org/sonar/java/checks/spring/SpringIncompatibleTransactionalCheck.java

```diff
@@ -0,0 +1,196 @@
+    BlockTree methodBody = method.block();
+    if (methodBody == null) {
+      return;
+    }
+    methodBody.accept(new BaseTreeVisitor() {
+      @Override
+      public void visitMethodInvocation(MethodInvocationTree methodInvocation) {
+        super.visitMethodInvocation(methodInvocation);
+        Symbol calleeMethodSymbol = methodInvocation.symbol();
+        if (methodsPropagationMap.containsKey(calleeMethodSymbol) && methodInvocationOnThisInstance(methodInvocation)) {
+          String calleePropagation = methodsPropagationMap.get(calleeMethodSymbol);
+          checkIncompatiblePropagation(methodInvocation, callerPropagation, calleeMethodSymbol, calleePropagation);
+        }
+      }
+    });
+  }
+
+  private static boolean methodInvocationOnThisInstance(MethodInvocationTree methodInvocation) {
+    ExpressionTree expression = methodInvocation.methodSelect();
+    if (expression.is(Tree.Kind.MEMBER_SELECT)) {
+      expression = ((MemberSelectExpressionTree) expression).expression();
+      return expression.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expression).name().equals("this");
+    }
+    return expression.is(Tree.Kind.IDENTIFIER);
+  }
+
+  private void checkIncompatiblePropagation(MethodInvocationTree methodInvocation, @Nullable String callerPropagation, Symbol calleeMethodSymbol, String calleePropagation) {
+    Set<String> incompatiblePropagation = INCOMPATIBLE_PROPAGATION_MAP.getOrDefault(callerPropagation, Collections.emptySet());
+    if (incompatiblePropagation.contains(calleePropagation)) {
+      String message = "\"" + calleeMethodSymbol.name() + "'s\" @Transactional requirement is incompatible with the one for this method.";
+      List<JavaFileScannerContext.Location> secondaryLocations = Collections.singletonList(
+        new JavaFileScannerContext.Location("", ((MethodTree) calleeMethodSymbol.declaration()).simpleName()));
+      reportIssue(ExpressionUtils.methodName(methodInvocation), message, secondaryLocations, null);
+    }
+  }
+
+  private static Map<Symbol, String> collectMethodsPropagation(ClassTree classTree) {
+    Map<Symbol, String> methodPropagationMap = new HashMap<>();
+    String classPropagation = getPropagation(classTree.symbol(), null);
+    for (Tree member : classTree.members()) {
+      if (member.is(Tree.Kind.METHOD)) {
+        MethodTree method = (MethodTree) member;
+        methodPropagationMap.put(method.symbol(), getPropagation(method.symbol(), classPropagation));
+      }
+    }
+    return methodPropagationMap;
+  }
+
+  private static boolean hasSameValues(Map<Symbol, String> methodsPropagationMap) {
+    if (methodsPropagationMap.size() < 2) {
+      return true;
+    }
+    String firstValue = methodsPropagationMap.values().iterator().next();
+    return methodsPropagationMap.values().stream().allMatch(value -> Objects.equals(value, firstValue));
+  }
+
+  private static boolean hasMethodsAnnotatedWithTransactional(ClassTree classTree) {
```

Comment:
This will also match fields annotated with the two annotation specified, no idea if you did that on purpose or not, but there is a mismatch between what the method is doing and its name.

Reply:
The method was just optimization, I removed it.